### ISSUE TITLE
Implementing GAN

### ISSUE THREAD

**Issue by webstorms** [deeplearning4j/deeplearning4j#3452]
I am trying to implement a vanilla GAN in dl4j where the generator (G) and discriminator (D) are both feedforward neural networks.

My first attempt was to try to implement a a customLoss function for both the D and G by implementing the ILossFunction interface as done in the dl4j examples. I got stuck on implementing the computeGradient method. I am not sure how I would determine the derivative d(D(x_i))/d(yhat) and d(D(G(z_i)))/d(yhat).

Would I better off trying to use the MultiLayerNetworkExternalErrors example?

Any help or comments would be appreciated.

**Comment by huitseeker** [comment]
![](https://github.trello.services/images/mini-trello-icon.png) [DataVec & al. - improve open-source documentation related to workflow](https://trello.com/c/3u3jt2lF/5-datavec-al-improve-open-source-documentation-related-to-workflow)

**Comment by wmeddie** [comment]
Here's my non-working attempt.  Didn't spend too long on it but maybe it might help.  I don't expect any good results until someone implements transposed convolution layers though.

https://gist.github.com/wmeddie/487becab1d4dbdb6d240744149112577

**Comment by webstorms** [comment]
Awesome! Thanks for sharing

**Comment by webstorms** [comment]
For the crash that happens when you call gen.fit(noise), what do those matrices represent [400, 32]x[784, 400]?

**Comment by webstorms** [comment]
In addition could you explain how you set up the loss functions? I don't see any log operators used.

**Comment by webstorms** [comment]
The exception can be avoided by changing the batch size of MnistDataSetIterator to 784, but then new exceptions pop up when calling dis.fit(ds);

**Comment by agibsonccc** [comment]
We will do a version of this once autodiff is finished. For now I'm closing this.

**Comment by webstorms** [comment]
When are you planning to release autodiff? 👍
